README.md
```

Library:
    bfix.hpp

A Generic API for Bit Manipulation in C/C++

    - insert(bfi()) or extract(bfx()) bit fields from an unsigned char array of arbitrary length

    - can handle 32 or 64 bit machines

    - big endian, little endian or run time checking

    - a C++ header file only library

    - simple library, two routines bfi()(insert bit field) and bfx()(extract bit field)

    - no library dependencies(except standard header files)

    - offset bit numbers are from 1(start of array) to unlimited

    - verification tests that show code usage examples, see bfix_test.cpp

    - see cut/paste Examples section and API(bfi(), bfx()) section below

    - a bfix-compile.tar file is provided with a non-header compiled(clang, clang++, gcc, g++) version

    - compiles with clang++/g++

    - see article in Embedded Systems Programming, Jul. 1999,
      "A Generic API for Bit Manipulation in C"(included)

    - this file is the entire documentation needed to use the library

    - Author: Richard Hogaboom, richard.hogaboom@gmail.com

Usage:

Files:

    bfix-compile.tar - non-header compiled(clang, clang++, gcc, g++) version
    bfix.doc         - documentation for library
    bfix.hpp         - header only C++ implementation
    bfix.mk          - build script, compiles/executes test code
    bfix_test.cpp    - test code

Examples:

API:

/*
 * File: bfix.hpp
 *
 *     Based on article in Embedded Systems Programming, Jul. 1999,
 *     "A Generic API for Bit Manipulation in C" by Richard Hogaboom
 *
 * Description:
 *
 *     bfi()/bfx() are used to insert and extract bit fields from an arbitrary length array
 *     of unsigned chars pointed to by an unsigned char* pointer.
 *
 * Functions:
 *
 *         int
 *     bfi
 *         (
 *             unsigned char *cptr,
 *             unsigned long bit_offset,
 *             unsigned long bit_len,
 *             long value,
 *             unsigned int endian
 *         )
 *         e.g. int return = bfi(cptr, bit_offset, bit_len, value, endian);
 *
 *
 *         long
 *     bfx
 *         (
 *             const unsigned char *cptr,
 *             unsigned long bit_offset,
 *             unsigned long bit_len,
 *             unsigned int endian
 *         )
 *         e.g. long value = bfx(cptr, bit_offset, bit_len, endian);
 *
 * Error Handling:
 *
 *     1. Exceptions - None
 *     2. Debugging  - const int DEBUG = true for debugging output
 *     3. Returns
 *
 *            bfi():
 *                bit_offset < 1     - error, return -1
 *                bit_len < 1        - error, return -2
 *                bit_len > too long - error, return -3
 *                endian not 0-2     - error, return -4
 *                return 0           - success
 *
 *            bfx():
 *                bit_offset < 1     - error, return -1
 *                bit_len < 1        - error, return -2
 *                bit_len > too long - error, return -3
 *                endian not 0-2     - error, return -4
 *                return value       - success
 *
 * Notes:
 *     1.  in the following notes any annotation of the form n/m means n for 32 bit systems and
 *         m for 64 bit systems.  operation on 32 or 64 bit systems should be transparent.
 *
 *     2.  bit_len should be <=32/64 to 25/57. it depends on the value of bit_offset.  the method
 *         always uses a memmove() of 4/8 bytes to a long temporary storage in which logical
 *         operations can be performed.  this means that the bit_len can be at most 4/8 bytes,
 *         but in a case in which the start bit is not at the beginning of a byte, then the
 *         bit_len can only extend until the end of the 4/8'th byte.  if the start bit is the
 *         last bit of a byte this will limit bit_len to 25/57 bits - the last bit of the first
 *         byte plus the next 3/7 bytes.
 *
 *     3.  4(32 bit machines)/8(64 bit machines) bytes are always read from the unsigned char
 *         array, modified and then written back.  this means that if you set the last bit of
 *         the array, then the next 3/7 bytes will be read and written back, thus seemingly
 *         overrunning the array.  if the 4/8 bytes does not overrun the array then no bits
 *         beyond the end of the array will be changed.  if the 4/8 bytes does overrun the
 *         array some provision must be made to deal with this possibility.  the array could
 *         be padded by 3/7 extra bytes.
 *
 *     4.  bit_offset+bit_len should not overrun the array.
 *
 *     5.  value should not be too long to fit into the bit field.  if it is, the high order bits
 *         in front of the low order bit_len bits will be truncated.
 *
 *     6.  all bit_len bits will be set and no bit outside the bit field will be changed.
 *
 *     7.  value may be negative and prefix 2's complement sign bits are truncated to fit into
 *         bit_len bits.
 *
 *     8.  use the lscpu cmd to determine 32/64 bit and endianness:
 *         $ lscpu
 *         Architecture:          x86_64
 *         CPU op-mode(s):        32-bit, 64-bit
 *         Byte Order:            Little Endian
 *
 *     9.  use lscpu to determine 32/64 bit machine and big or little endian
 *
 *     10. since bit_offset is unsigned long it cannot have negative values - however no
 *         check for too large is done
 *
 * Author: Richard Hogaboom
 *         richard.hogaboom@gmail.com
 *
 */

```


HISTORY.md
```

 bfix-1.0.7.tar.gz -
     a. doc changes

 bfix-1.0.6.tar.gz -
     a. minor code changes
     b. remove LLVM 4.0.0 scan-build program used for static analysis
     c. improve doc significantly

 bfix-1.0.5.tar.gz -
     a. minor code and doc changes

 bfix-1.0.4.tar.gz -
     a. improve test suite - add tests for all error conditions
     b. cerr -> fprintf() to restore compilation with gcc/clang
     c. the LLVM 4.0.0 scan-build program is used for static analysis
     d. no language dialects - too many incompatable combinations 
     e. many doc and minor code changes to bfix.cpp

 bfix-1.0.3.tar.gz -
     a. API change - add endian argument and get ride of endian defines
     b. do error checking all the time
     c. return more meaningful error codes
     d. do printing only on DEBUG = true
     e. for design/development use clang++(5.0) - verify with g++ on release
     f. use LLVM scan-build for static analysis
     g. fprintf() -> cerr

 bfix-1.0.2.tar.gz -
     a. add HISTORY.md file
     b. update license to Boost Software License, Version 1.0.
     c. remove bfix.README file(incorporated into README.md)
     d. rewrite README.md
     e. coding format changes - mostly doc
     f. fix g++ warning about possible uninitialized variable -
        some unneeded code - not a bug
     g. modified bfix.mk build script:
            a. add clean target
            b. add compile with gcc/g++/clang/clang++ option
            c. add language dialect option
            d. add debug/optimize option
     h. eliminated the C version.  the C++ version will now compile with
         gcc/g++/clang/clang++ - older versions of the compilers would not.
     i. eliminate .h headers from the C++ version
     j. verify code correct operation with latest:
        1. gcc/g++       - gcc (GCC) 6.3.1 20161221 (Red Hat 6.3.1-1)
        2. clang/clang++ - clang version 3.8.1 (tags/RELEASE_381/final)
     k. change error return for bfx() - was 0 on error - now is -1 on error.
        this was a bug - 0 return is a valid value - rarely seen because
        valid 0 value returns didn't cause errors

 bfix-1.0.1.tar.gz -
     a. delete bfix.tar.xz file

 bfix-1.0.0.tar.gz -
     a. initial versioned release - the code has been on GitHub for some time without
        versioning.  in future, releases will be versioned.
     b. see article at www.embedded.com:
        http://www.embedded.com/design/other/4219600/A-Generic-API-for-Bit-Manipulation-in-C

```


bfix/bfix_test.cpp

/*
 * test and example code for bfi()/bfx() bit field insert/extract library
 */

#include <string>
#include <cstdio>
#include <cstdlib>

#include "bfix.hpp"


    int
main()
{
    long bitfield1, bitfield2, bitfield3, bitfield4, bitfield5;

    int BYTES_PER_LONG = sizeof(unsigned long);

    /* 16 bytes unsigned char array plus 7 bytes pad(3 needed for 32 bit systems and 7 needed for 64 bit systems) */
    unsigned char c[16+7];


    fprintf(stderr, "\nTest and example code for bfix(bfi()/bfx()) bit field insert/extract project and libraries\n\n");
    fprintf(stderr, "BYTES_PER_LONG = %d\n\n\n", BYTES_PER_LONG);

    for ( int i = 0 ; i < 16+7 ; i++ )
    {
        c[i]  = 0x00;
    }

    bfi(c, 1, 1, 1, 0);
    bfi(c, 5, 2, 3, 0);
    bfi(c, 17, 4, 15, 0);
    bfi(c, 97, 32, 0xffffffff, 0);
    bfi(c, 49, 8, 0xee, 0);
    bfi(c, 8, 8, 0xff, 0);

    fprintf(stderr, "Tests on 32 and 64 bit systems.\n\n");
    fprintf(stderr, "c(should be) = 8d fe f0 00 00 00 ee 00 00 00 00 00 ff ff ff ff\n");

    fprintf(stderr, "           c = %02x %02x %02x %02x %02x %02x %02x %02x ",
        c[0], c[1], c[2], c[3], c[4], c[5], c[6], c[7]);
    fprintf(stderr, "%02x %02x %02x %02x %02x %02x %02x %02x\n",
        c[8], c[9], c[10], c[11], c[12], c[13], c[14], c[15]);

    bitfield1 = bfx(c, 1, 1, 0);
    bitfield2 = bfx(c, 5, 2, 0);
    bitfield3 = bfx(c, 17, 4, 0);
    bitfield4 = bfx(c, 113, 16, 0);
    bitfield5 = bfx(c, 49, 8, 0);

    fprintf(stderr,
        "bitfield1(should be 1) = %ld\n"
        "bitfield2(should be 3) = %ld\n"
        "bitfield3(should be 15) = %ld\n"
        "bitfield4(should be 65535) = %ld\n"
        "bitfield5(should be 238) = %ld\n",
        bitfield1, bitfield2, bitfield3, bitfield4, bitfield5);

    if ( BYTES_PER_LONG == 8 )
    {
        unsigned long ff, ee;

        fprintf(stderr, "\n\nTests on 64 bit systems.\n\n");
        for ( int i = 0 ; i < 16+7 ; i++ )
        {
            c[i]  = 0x00;
        }

        for ( int i = 0 ; i < BYTES_PER_LONG ; i++ )
        {
            ((unsigned char *)&ff)[i] = 0xff;
            ((unsigned char *)&ee)[i] = 0xee;
        }

        bfi(c, 1, 64, ff, 0);
        bfi(c, 65, 64, ee, 0);

        fprintf(stderr, "c(should be) = ff ff ff ff ff ff ff ff ee ee ee ee ee ee ee ee\n");
        fprintf(stderr, "           c = %02x %02x %02x %02x %02x %02x %02x %02x ",
            c[0], c[1], c[2], c[3], c[4], c[5], c[6], c[7]);
        fprintf(stderr, "%02x %02x %02x %02x %02x %02x %02x %02x\n",
            c[8], c[9], c[10], c[11], c[12], c[13], c[14], c[15]);

        bitfield1 = bfx(c, 1, 64, 0);
        bitfield2 = bfx(c, 65, 64, 0);
        bitfield3 = bfx(c, 33, 64, 0);

        fprintf(stderr,
            "bitfield1(should be ffffffffffffffff) = %016lx\n"
            "bitfield2(should be eeeeeeeeeeeeeeee) = %016lx\n"
            "bitfield3(should be ffffffffeeeeeeee) = %016lx\n",
            bitfield1, bitfield2, bitfield3);
    }

    fprintf(stderr, "\n\nError tests.\n\n");

    int ret;

    ret = bfi(c, 1, 1, 1, 0);
    fprintf(stderr, "bfi(): Set bit_offset = 1(return should be 0) = %d\n", ret);

    ret = bfi(c, 0, 1, 1, 0);
    fprintf(stderr, "bfi(): Set bit_offset = 0(return should be -1) = %d\n", ret);

    ret = bfi(c, 1, 0, 1, 0);
    fprintf(stderr, "bfi(): Set bit_len = 0,(return should be -2) = %d\n", ret);

    ret = bfi(c, 1, 65, 1, 0);
    fprintf(stderr, "bfi(): Set bit_len = 65,(return should be -3) = %d\n", ret);

    ret = bfi(c, 1, 1, 1, 3);
    fprintf(stderr, "bfi(): Set endian = 3,(return should be -4) = %d\n", ret);

    fprintf(stderr, "\n");

    ret = bfx(c, 1, 1, 0);
    fprintf(stderr, "bfx(): Set bit_offset = 1(return should be 1) = %d\n", ret);

    ret = bfx(c, 0, 1, 0);
    fprintf(stderr, "bfx(): Set bit_offset = 0(return should be -1) = %d\n", ret);

    ret = bfx(c, 1, 0, 0);
    fprintf(stderr, "bfx(): Set bit_len = 0,(return should be -2) = %d\n", ret);

    ret = bfx(c, 1, 65, 0);
    fprintf(stderr, "bfx(): Set bit_len = 65,(return should be -3) = %d\n", ret);

    ret = bfx(c, 1, 1, 3);
    fprintf(stderr, "bfx(): Set endian = 3,(return should be -4) = %d\n", ret);

    exit(0);
}



bfix/c/bfix_test.c
/*
 * test program for bfi() and bfx()
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "bfix.h"


   int
main()
{
   unsigned long i, f1, f2, f3, f4, f5;
   unsigned int BYTES_PER_LONG;

   /* 16 bytes unsigned char array plus 7 bytes pad(3 needed for 32 bit systems and 7 needed for 64 bit systems)*/
   unsigned char c[16+7];


   for (i=0 ; i<16+7 ; i++)
   {
      c[i]  = 0x00;
   }

   f1 = f2 = f3 = f4 = f5 = 0;

   bfi(c, 1, 1, 1);
   bfi(c, 5, 2, 3);
   bfi(c, 17, 4, 15);
   bfi(c, 97, 32, 0xffffffff);
   bfi(c, 49, 8, 0xee);
   bfi(c, 8, 8, 0xff);

   fprintf(stderr, "Tests on 32 and 64 bit systems.\n");
   fprintf(stderr, "c(should be) = 8d fe f0 00 00 00 ee 00 00 00 00 00 ff ff ff ff\n");

   fprintf(stderr, "           c = %02x %02x %02x %02x %02x %02x %02x %02x ",
      c[0], c[1], c[2], c[3], c[4], c[5], c[6], c[7]);
   fprintf(stderr, "%02x %02x %02x %02x %02x %02x %02x %02x\n",
      c[8], c[9], c[10], c[11], c[12], c[13], c[14], c[15]);

   f1 = bfx(c, 1, 1);
   f2 = bfx(c, 5, 2);
   f3 = bfx(c, 17, 4);
   f4 = bfx(c, 113, 16);
   f5 = bfx(c, 49, 8);

   fprintf(stderr,
      "f1(should be 1) = %ld\n"
      "f2(should be 3) = %ld\n"
      "f3(should be 15) = %ld\n"
      "f4(should be 65535) = %ld\n"
      "f5(should be 238) = %ld\n",
      f1, f2, f3, f4, f5);

   BYTES_PER_LONG = sizeof(unsigned long);
   if (BYTES_PER_LONG == 8)
   {
      unsigned long ff, ee;


      fprintf(stderr, "\n\nTests on 64 bit systems.\n");
      for (i=0 ; i<16+7 ; i++)
      {
         c[i]  = 0x00;
      }

      for (i=0 ; i<BYTES_PER_LONG ; i++)
      {
         ((unsigned char *)&ff)[i] = 0xff;
         ((unsigned char *)&ee)[i] = 0xee;
      }

      bfi(c, 1, 64, ff);
      bfi(c, 65, 64, ee);

      fprintf(stderr, "c(should be) = ff ff ff ff ff ff ff ff ee ee ee ee ee ee ee ee\n");
      fprintf(stderr, "           c = %02x %02x %02x %02x %02x %02x %02x %02x ",
         c[0], c[1], c[2], c[3], c[4], c[5], c[6], c[7]);
      fprintf(stderr, "%02x %02x %02x %02x %02x %02x %02x %02x\n",
         c[8], c[9], c[10], c[11], c[12], c[13], c[14], c[15]);

      f1 = bfx(c, 1, 64);
      f2 = bfx(c, 65, 64);
      f3 = bfx(c, 33, 64);

      fprintf(stderr,
         "f1(should be ffffffffffffffff) = %016lx\n"
         "f2(should be eeeeeeeeeeeeeeee) = %016lx\n"
         "f3(should be ffffffffeeeeeeee) = %016lx\n",
         f1, f2, f3);
   }

   exit(0);
}



bfix/cpp/bfix_test.cpp
/*
 * test program for bfi() and bfx()
 */

#include <cstring>
#include <cstdio>
#include <cstdlib>
#include <iostream>

#include "bfix.hpp"


   int
main()
{
   long i, f1, f2, f3, f4, f5;
   unsigned int BYTES_PER_LONG;

   /* 16 bytes unsigned char array plus 7 bytes pad(3 needed for 32 bit systems and 7 needed for 64 bit systems)*/
   unsigned char c[16+7];


   for (i=0 ; i<16+7 ; i++)
   {
      c[i]  = 0x00;
   }

   f1 = f2 = f3 = f4 = f5 = 0;

   bfi(c, 1, 1, 1, 0);
   bfi(c, 5, 2, 3, 0);
   bfi(c, 17, 4, 15, 0);
   bfi(c, 97, 32, 0xffffffff, 0);
   bfi(c, 49, 8, 0xee, 0);
   bfi(c, 8, 8, 0xff, 0);

   fprintf(stderr, "Tests on 32 and 64 bit systems.\n");
   fprintf(stderr, "c(should be) = 8d fe f0 00 00 00 ee 00 00 00 00 00 ff ff ff ff\n");

   fprintf(stderr, "           c = %02x %02x %02x %02x %02x %02x %02x %02x ",
      c[0], c[1], c[2], c[3], c[4], c[5], c[6], c[7]);
   fprintf(stderr, "%02x %02x %02x %02x %02x %02x %02x %02x\n",
      c[8], c[9], c[10], c[11], c[12], c[13], c[14], c[15]);

   f1 = bfx(c, 1, 1, 0);
   f2 = bfx(c, 5, 2, 0);
   f3 = bfx(c, 17, 4, 0);
   f4 = bfx(c, 113, 16, 0);
   f5 = bfx(c, 49, 8, 0);

   fprintf(stderr,
      "f1(should be 1) = %ld\n"
      "f2(should be 3) = %ld\n"
      "f3(should be 15) = %ld\n"
      "f4(should be 65535) = %ld\n"
      "f5(should be 238) = %ld\n",
      f1, f2, f3, f4, f5);

   BYTES_PER_LONG = sizeof(unsigned long);
   if (BYTES_PER_LONG == 8)
   {
      unsigned long ff, ee;


      fprintf(stderr, "\n\nTests on 64 bit systems.\n");
      for (i=0 ; i<16+7 ; i++)
      {
         c[i]  = 0x00;
      }

      for (i=0 ; i<BYTES_PER_LONG ; i++)
      {
         ((unsigned char *)&ff)[i] = 0xff;
         ((unsigned char *)&ee)[i] = 0xee;
      }

      bfi(c, 1, 64, ff, 0);
      bfi(c, 65, 64, ee, 0);

      fprintf(stderr, "c(should be) = ff ff ff ff ff ff ff ff ee ee ee ee ee ee ee ee\n");
      fprintf(stderr, "           c = %02x %02x %02x %02x %02x %02x %02x %02x ",
         c[0], c[1], c[2], c[3], c[4], c[5], c[6], c[7]);
      fprintf(stderr, "%02x %02x %02x %02x %02x %02x %02x %02x\n",
         c[8], c[9], c[10], c[11], c[12], c[13], c[14], c[15]);

      f1 = bfx(c, 1, 64, 0);
      f2 = bfx(c, 65, 64, 0);
      f3 = bfx(c, 33, 64, 0);

      fprintf(stderr,
         "f1(should be ffffffffffffffff) = %016lx\n"
         "f2(should be eeeeeeeeeeeeeeee) = %016lx\n"
         "f3(should be ffffffffeeeeeeee) = %016lx\n",
         f1, f2, f3);
   }

   exit(0);
}



bfix/cpp/bfix.cpp

//
//          Copyright Richard Allen Hogaboom 1997 - 2017.
// Distributed under the Boost Software License, Version 1.0.
//    (See accompanying file LICENSE_1_0.txt or copy at
//          http://www.boost.org/LICENSE_1_0.txt)
//
// Boost Software License - Version 1.0 - August 17th, 2003
//
// Permission is hereby granted, free of charge, to any person or organization
// obtaining a copy of the software and accompanying documentation covered by
// this license (the "Software") to use, reproduce, display, distribute,
// execute, and transmit the Software, and to prepare derivative works of the
// Software, and to permit third-parties to whom the Software is furnished to
// do so, all subject to the following:
//
// The copyright notices in the Software and this entire statement, including
// the above license grant, this restriction and the following disclaimer,
// must be included in all copies of the Software, in whole or in part, and
// all derivative works of the Software, unless such copies or derivative
// works are solely in the form of machine-executable object code generated by
// a source language processor.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT. IN NO EVENT
// SHALL THE COPYRIGHT HOLDERS OR ANYONE DISTRIBUTING THE SOFTWARE BE LIABLE
// FOR ANY DAMAGES OR OTHER LIABILITY, WHETHER IN CONTRACT, TORT OR OTHERWISE,
// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.
//

/*
 *==================================================================================================
 *
 * File Name:
 *     bfix.cpp
 *
 *==================================================================================================
 *
 * Functions:
 *
 *         int
 *     bfi
 *         (
 *             unsigned char *cptr,
 *             unsigned long bit_offset,
 *             unsigned long bit_len,
 *             long value,
 *             unsigned int endian
 *         )
 *         e.g. int return = bfi(cptr, bit_offset, bit_len, value, endian);
 *
 *
 *         long
 *     bfx
 *         (
 *             const unsigned char *cptr,
 *             unsigned long bit_offset,
 *             unsigned long bit_len,
 *             unsigned int endian
 *         )
 *         e.g. long value = bfx(cptr, bit_offset, bit_len, endian);
 *
 *==================================================================================================
 *
 * Description:
 *
 *     bfi()/bfx() are used to insert and extract bit fields from an arbitrary length array
 *     of unsigned chars pointed to by an unsigned char* pointer.
 *
 *==================================================================================================
 *
 * Error Handling
 *
 *     1. Exceptions - None
 *     2. Debugging  - const int DEBUG = true for debugging output
 *     3. Returns
 *
 *            bfi():
 *                bit_offset < 1     - error, return -1
 *                bit_len < 1        - error, return -2
 *                bit_len > too long - error, return -3
 *                endian not 0-2     - error, return -4
 *                return 0           - success
 *
 *            bfx():
 *                bit_offset < 1     - error, return -1
 *                bit_len < 1        - error, return -2
 *                bit_len > too long - error, return -3
 *                endian not 0-2     - error, return -4
 *                return value       - success
 *
 *==================================================================================================
 *
 * Notes:
 *     1. in the following notes any annotation of the form n/m means n for 32 bit systems and
 *        m for 64 bit systems.  operation on 32 or 64 bit systems should be transparent.
 *
 *     2. bit_len should be <=32/64 to 25/57. it depends on the value of bit_offset.  the method
 *        always uses a memmove() of 4/8 bytes to a long temporary storage in which logical
 *        operations can be performed.  this means that the bit_len can be at most 4/8 bytes,
 *        but in a case in which the start bit is not at the beginning of a byte, then the
 *        bit_len can only extend until the end of the 4/8'th byte.  if the start bit is the
 *        last bit of a byte this will limit bit_len to 25/57 bits - the last bit of the first
 *        byte plus the next 3/7 bytes.
 *
 *     3. 4(32 bit machines)/8(64 bit machines) bytes are always read from the unsigned char
 *        array, modified and then written back.  this means that if you set the last bit of
 *        the array, then the next 3/7 bytes will be read and written back, thus seemingly
 *        overrunning the array.  if the 4/8 bytes does not overrun the array then no bits
 *        beyond the end of the array will be changed.  if the 4/8 bytes does overrun the
 *        array some provision must be made to deal with this possibility.  the array could
 *        be padded by 3/7 extra bytes.
 *
 *     4. bit_offset+bit_len should not overrun the array.
 *
 *     5. value should not be too long to fit into the bit field.  if it is, the high order bits
 *        in front of the low order bit_len bits will be truncated.
 *
 *     6. all bit_len bits will be set and no bit outside the bit field will be changed.
 *
 *     7. value may be negative and prefix 2's complement sign bits are truncated to fit into
 *        bit_len bits.
 *
 *     8. use the lscpu cmd to determine 32/64 bit and endianness:
 *        $ lscpu
 *        Architecture:          x86_64
 *        CPU op-mode(s):        32-bit, 64-bit
 *        Byte Order:            Little Endian
 *
 *==================================================================================================
 *
 * Author: Richard Hogaboom
 *         richard.hogaboom@gmail.com
 *
 *==================================================================================================
 */

#include <cstring>
#include <cstdio>

#include "bfix.hpp"

const bool DEBUG = false;

/*
 *==================================================================================================
 *     bfi()
 *
 * Purpose: 
 *     extract bit field from an array of chars pointed to by an unsigned char* pointer
 *
 * Usage:
 *     unsigned char *cptr
 *     unsigned long bit_offset
 *     unsigned long bit_len
 *     long value
 *     unsigned int endian
 *
 *     int return = bfi(cptr, bit_offset, bit_len, value, endian);
 *
 * Returns:
 *     bit_offset < 1     - error, return -1
 *     bit_len < 1        - error, return -2
 *     bit_len > too long - error, return -3
 *     endian not 0-2     - error, return -4
 *     return 0           - success
 *
 * Parameters:
 *     unsigned char *cptr      - pointer to unsigned char array
 *     unsigned long bit_offset - bit offset(starting from 1) from the start of the char array
 *     unsigned long bit_len    - bit length of field to insert
 *     long value               - value to insert
 *     unsigned int endian      - endian enum, 0(run time checking), 1(big endian), 2(little endian)
 *
 * Comments:
 *     4(32 bit machines)/8(64 bit machines) bytes are always read from the unsigned char
 *     array, modified and then written back.  this means that if you set the last bit of
 *     the array, then the next 3/7 bytes will be read and written back, thus seemingly
 *     overrunning the array.  if the 4/8 bytes does not overrun the array then no bits
 *     beyond the end of the array will be changed.  if the 4/8 bytes does overrun the
 *     array some provision must be made to deal with this possibility.  the array could
 *     be padded by 3/7 extra bytes.
 */

    int
bfi
    (
        unsigned char *cptr,
        unsigned long bit_offset,
        unsigned long bit_len,
        long value,
        unsigned int endian
    )
{
    /* machine dependencies */
    const unsigned int BITS_PER_BYTE = 8;
    unsigned int BYTES_PER_LONG = sizeof(unsigned long);
    unsigned int BITS_PER_LONG = BYTES_PER_LONG * BITS_PER_BYTE;

    unsigned long l, m;
    unsigned int i, j, size;
    unsigned char* c = (unsigned char*)&l;
    unsigned char tmp;

    unsigned long mask;
    unsigned long byte_offset;
    unsigned long pre_shift;
    unsigned long post_shift;


    for ( i=0 ; i < BYTES_PER_LONG ; i++ )
    {
        ((unsigned char *)&mask)[i] = 0xff;
    }

    if ( bit_offset < 1 )
    {
        if ( DEBUG == true )
        {
            fprintf(stderr, "bfi: bit_offset = %ld is < 1.\n", bit_offset);
        }
        return -1;
    }

    if ( bit_len < 1 )
    {
        if ( DEBUG == true )
        {
            fprintf(stderr, "bfi: bit_len = %ld is < 1.\n", bit_len);
        }
        return -2;
    }

    /*
     * calculate byte offset(first byte=0) of start of
     * BYTES_PER_LONG bytes containing bit field
     */
    byte_offset = (bit_offset-1)/BITS_PER_BYTE;

    /*
     * calculate how many bits to shift bit field left
     * to clear bits above bit field
     */
    pre_shift = bit_offset - byte_offset*BITS_PER_BYTE - 1;

    if ( bit_len > (BITS_PER_LONG-pre_shift) )
    {
        if ( DEBUG == true )
        {
            fprintf(stderr, "bfi: bit_len = %ld is too long.\n", bit_len);
        }
        return -3;
    }

    /*
     * calculate how many bits to shift bit field left
     * to clear bits below bit field
     */
    post_shift = BITS_PER_LONG - bit_len - pre_shift;

    /*
     * move bit field into position over bits to set in l
     * corresponding to correct bits in cptr
     */
    value <<= post_shift;

    /* zero out mask bits after bit field */
    mask <<= post_shift;

    /* zero out mask bits before bit field */
    mask <<= pre_shift;
    mask >>= pre_shift;

    /* zero out value bits before and after bit field */
    value &= mask;

    /* move BYTES_PER_LONG bytes to tmp storage */
    memmove((unsigned char *)&l, &cptr[byte_offset], BYTES_PER_LONG);

    switch ( endian )
    {
        case 1:
            /* big endian */
            break;
        case 2:
            /* little endian */
            size = sizeof(l);
            for ( i = 0 ; i < size/2; i++ )
            {
                j = size - i - 1;
                tmp = c[i];
                c[i] = c[j];
                c[j] = tmp;
            }
            break;
        case 0:
            /* run time checking */
            m = l;
            l = 1;
            if (c[0])
            {
                /* little endian */
                l = m;
                size = sizeof(l);
                for ( i = 0 ; i < size/2; i++ )
                {
                    j = size - i - 1;
                    tmp = c[i];
                    c[i] = c[j];
                    c[j] = tmp;
                }
            }
            else
            {
                /* big endian */
                l = m;
            }
            break;

        default:
            if ( DEBUG == true )
            {
                fprintf(stderr, "bfi: endian = %d not 0-2.\n", endian);
            }
            return -4;
            break;
    }

    /* zero out bit field bits and then or value bits into them */
    l = (l & (~mask)) | value;

    switch ( endian )
    {
        case 1:
            /* big endian */
            break;
        case 2:
            /* little endian */
            size = sizeof(l);
            for ( i = 0 ; i < size/2; i++ )
            {
                j = size - i - 1;
                tmp = c[i];
                c[i] = c[j];
                c[j] = tmp;
            }
            break;
        case 0:
            /* run time checking */
            m = l;
            l = 1;
            if (c[0])
            {
                /* little endian */
                l = m;
                size = sizeof(l);
                for ( i = 0 ; i < size/2; i++ )
                {
                    j = size - i - 1;
                    tmp = c[i];
                    c[i] = c[j];
                    c[j] = tmp;
                }
            }
            else
            {
                /* big endian */
                l = m;
            }
            break;

        default:
            break;
    }

    /* move tmp storage back to cptr array */
    memmove(&cptr[byte_offset], (unsigned char *)&l, BYTES_PER_LONG);

    return 0;
}


/*
 *==================================================================================================
 *     bfx()
 *
 * Purpose: 
 *     extract bit field from an array of chars pointed to by an unsigned char* pointer
 *
 * Usage:
 *     const unsigned char *cptr
 *     unsigned long bit_offset
 *     unsigned long bit_len
 *     unsigned int endian
 *
 *     long value = bfx(cptr, bit_offset, bit_len, endian);
 *
 * Returns:
 *     bit_offset < 1     - error, return -1
 *     bit_len < 1        - error, return -2
 *     bit_len > too long - error, return -3
 *     endian not 0-2     - error, return -4
 *     return value       - success
 *
 * Parameters:
 *     const unsigned char *cptr - const pointer to unsigned char array
 *     unsigned long bit_offset  - bit offset(starting from 1) from the start of the char array
 *     unsigned long bit_len     - bit length of field to extract
 *     unsigned int endian       - endian enum, 0(run time checking), 1(big endian), 2(little endian)
 *
 * Comments:
 */

    long
bfx
    (
        const unsigned char *cptr,
        unsigned long bit_offset,
        unsigned long bit_len,
        unsigned int endian
    )
{
    /* machine dependencies */
    const unsigned int BITS_PER_BYTE = 8;
    unsigned int BYTES_PER_LONG = sizeof(unsigned long);
    unsigned int BITS_PER_LONG = BYTES_PER_LONG * BITS_PER_BYTE;

    unsigned long l;
    unsigned int i, j, size;
    unsigned char* c = (unsigned char*)&l;
    unsigned char tmp;

    unsigned long byte_offset;
    unsigned long left_shift;
    unsigned long right_shift;


    if ( bit_offset < 1 )
    {
        if ( DEBUG == true )
        {
            fprintf(stderr, "bfi: bit_offset = %ld is < 1.\n", bit_offset);
        }
        return -1;
    }

    if ( bit_len < 1 )
    {
        if ( DEBUG == true )
        {
            fprintf(stderr, "bfi: bit_len = %ld is < 1.\n", bit_len);
        }
        return -2;
    }

    /*
     * calculate byte offset(first byte=0) of start of
     * BYTES_PER_LONG bytes containing bit field
     */
    byte_offset = (bit_offset-1)/BITS_PER_BYTE;

    /*
     * calculate how many bits to shift bit field left
     * to clear bits above bit field
     */
    left_shift = bit_offset - byte_offset*BITS_PER_BYTE - 1;

    if ( bit_len > (BITS_PER_LONG-left_shift) )
    {
        if ( DEBUG == true )
        {
            fprintf(stderr, "bfi: bit_len = %ld is too long.\n", bit_len);
        }
        return -3;
    }

    /*
     * calculate how many bits to shift bit field right
     * to right justify bit field
     */
    right_shift = BITS_PER_LONG - bit_len;

    switch ( endian )
    {
        case 1:
            /* big endian */
            /* move BYTES_PER_LONG bytes to tmp storage */
            memmove((unsigned char *)&l, &cptr[byte_offset], BYTES_PER_LONG);
            break;

        case 2:
            /* little endian */
            memmove((unsigned char *)&l, &cptr[byte_offset], BYTES_PER_LONG);
            size = sizeof(l);
            for ( i = 0 ; i < size/2; i++ )
            {
                j = size - i - 1;
                tmp = c[i];
                c[i] = c[j];
                c[j] = tmp;
            }
            break;

        case 0:
            /* run time checking */
            l = 1;
            if (c[0])
            {
                /* little endian */
                memmove((unsigned char *)&l, &cptr[byte_offset], BYTES_PER_LONG);
                size = sizeof(l);
                for ( i = 0 ; i < size/2; i++ )
                {
                    j = size - i - 1;
                    tmp = c[i];
                    c[i] = c[j];
                    c[j] = tmp;
                }
            }
            else
            {
                /* big endian */
                memmove((unsigned char *)&l, &cptr[byte_offset], BYTES_PER_LONG);
            }
            break;

        default:
            if ( DEBUG == true )
            {
                fprintf(stderr, "bfx: endian = %d not 0-2.\n", endian);
            }
            return -4;
            break;
    }

    /*
     * clear bits above bit field, right justify bit
     * field, and return this value
     */
    return ( (l << left_shift) >> right_shift );
}

